python/hail/methods/statgen.py

```diff
@@ -126,3 +129,191 @@ def ld_matrix(dataset, force_local=False):
 
     jldm = Env.hail().methods.LDMatrix.apply(dataset._jvds, force_local)
     return LDMatrix(jldm)
+
+
+@handle_py4j
+@require_biallelic
+@typecheck(dataset=MatrixTable,
+           k=integral,
+           compute_loadings=bool,
+           as_array=bool)
+def hwe_normalized_pca(dataset, k=10, compute_loadings=False, as_array=False):
+    """Run principal component analysis (PCA) on the Hardy-Weinberg-normalized call matrix.
+
+    Examples
+    --------
+
+    .. testsetup::
+
+        import hail.methods as methods
+
+    >>> eigenvalues, scores, loadings = methods.hwe_normalized_pca(dataset, k=15)
+
+    Notes
+    -----
+    Variants that are all homozygous reference or all homozygous variant are removed before evaluation.
+
+    Parameters
+    ----------
+    dataset : :class:`MatrixTable`
+        Dataset.
+    k : :obj:`int`
+        Number of principal components.
+    compute_loadings : :obj:`bool`
+        If ``True``, compute row loadings.
+    as_array : :obj:`bool`
+        If ``True``, return scores and loadings as an array field. If ``False``, return
+        one field per element (`PC1`, `PC2`, ... `PCk`).
+
+    Returns
+    -------
+    (:obj:`list` of :obj:`float`, :class:`Table`, :class:`Table`)
+        List of eigenvalues, table with column scores, table with row loadings.
+    """
+
+    dataset = dataset.annotate_rows(AC = agg.sum(dataset.GT.num_alt_alleles()),
+                                    n_called = agg.count_where(functions.is_defined(dataset.GT)))
+    dataset = dataset.filter_rows((dataset.AC > 0) & (dataset.AC < 2 * dataset.n_called)).persist()
+
+    n_variants = dataset.count_rows()
+    if n_variants == 0:
+        raise FatalError(
+            "Cannot run PCA: found 0 variants after filtering out monomorphic sites.")
+    info("Running PCA using {} variants.".format(n_variants))
+
+    # FIXME: use bind
+    mean_gt = dataset.AC / dataset.n_called
+    entry_expr = functions.cond(functions.is_defined(dataset.GT),
+                                (dataset.GT.num_alt_alleles() - mean_gt) /
+                                functions.sqrt(mean_gt * (2 - mean_gt) * n_variants / 2),
+                                0)
+    result = pca(dataset,
+                 entry_expr,
+                 k,
+                 compute_loadings,
+                 as_array)
+    dataset.unpersist()
+    return result
+
+
+@handle_py4j
+@typecheck_method(entry_expr=Expression,
+                  k=integral,
+                  compute_loadings=bool,
+                  as_array=bool)
+def pca(dataset, entry_expr, k=10, compute_loadings=False, as_array=False):
```

Comment:
Is the dataset redundant since entry_expr has to be row- and column-indexed expression and you can just use the source?

Reply:
Also, might be nice to accept a MatrixTable itself as the entry_expr to handle the case it has a single, numeric-convertable entry.

Reply:
I thought about this. It probably is redundant, but didn't want to confuse people. OK to remove the first argument?

Reply:
@konradjk and I both prefer it.  Who else might have an opinion?